src/basicfuns.jl

```diff
@@ -190,12 +190,12 @@ function logsumexp(X)
     isempty(X) && return log(sum(X))
     reduce(logaddexp, X)
 end
-function logsumexp(X::AbstractArray{T}) where {T<:Real}
+function logsumexp(X::AbstractArray{T}; dims=:) where {T<:Real}
     isempty(X) && return log(zero(T))
```

Comment:
This won't be correct when `dims != :`, right? At this point, probably better add an internal `_logsumexp(X, dims)` function dispatching on `dims::Colon` vs. `dims::Any`.

Reply:
Something like this:
```julia
logsumexp(X::AbstractArray{<:Real}; dims=:) = _logsumexp(X, dims)
_logsumexp(X::AbstractArray{<:Real}, dims::Colon) = ...
_logsumexp(X::AbstractArray{<:Real}, dims::Any) = ...
```

That's the pattern used in Base for most reduction functions.

Reply:
Actually the implementation works for `dims != :`. Things like `logsumexp(x; dims=[1,2])` also work. Do I still need this dispatch pattern here?

Reply:
Well, returning `log(zero(T))` clearly isn't correct in that case...

Reply:
Oh I see what you are referring to! I misunderstood your point. 
Do we want to only do `isempty(X) && return log(zero(T))` for `dims::Colon`?

Reply:
You need to handle that case also for other `dims`, since `maximum` throws an error for empty arrays. Or maybe a simpler solution is to use `reduce(max, X, dims=dims, init=log(zero(T)))` instead.

Reply:
I tried your solution.
```julia
function logsumexp(X::AbstractArray{T}; dims=:) where {T<:Real}
    isempty(X) && return reduce(max, X, dims=dims, init=log(zero(T)))
    u = maximum(X; dims=dims)
    u isa AbstractArray || isfinite(u) || return float(u)
    let u=u # avoid https://github.com/JuliaLang/julia/issues/15276
        u .+ log.(sum(exp.(X .- u); dims=dims))
    end
end
```
gives
```julia
julia> X = zeros(0, 0)
0×0 Array{Float64,2}

julia> logsumexp(X)
-Inf

julia> logsumexp(X; dims=1)
1×0 Array{Float64,2}

julia> logsumexp(X; dims=[1,2])
1×1 Array{Float64,2}:
 -Inf
```

Is this something we want?

Reply:
I guess, yes. That sounds consistent with what `sum` does.

But my suggestion was to use `reduce(max, X, dims=dims, init=log(zero(T)))` instead of `maximum`. It wouldn't make sense to use it only when `X` is empty.

Reply:
You mean replace both them empty check line and the `u = maximum...` with `reduce` right?

Reply:
Right. Can you add tests with `dims=1`, `dims=2` and `dims=[1, 2]` when the output isn't finite? Otherwise that case isn't covered.

Reply:
Test for `dims=1` was there. I just added tests for `dims=2` and `dims=[1,2]` as well.